### PR TITLE
[serve] Extend configurable HTTP options

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Currently, users can configure the host and port of the HTTP Proxy through the config. This extends `ServeDeploySchema` by adding the options `root_path` and `proxy_location`. This makes them configurable. It also changes the structure of `ServeDeploySchema` a bit to make it more future-proof in case we want to support more transport protocols.

E.g.:
```
proxy_location: HeadOnly
http_options:
    host: 127.0.0.1
    port: 8000
    root_path: /serve

applications:
  - name: "app1"
    ...
```

Additional things that are added in this PR:
- Added validator in `ServeDeploySchema` that prevents users from setting host and port in individual application configs. Originally they were just being ignored, so adding a validator to explicitly ban it is more clear for users.
  - removed `ServeApplicationSchema.to_deploy_schema` which was previously used in `controller.deploy_apps`
  - removed `test_deploy_http_override` as that test case is now not allowed.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@zcin can we make this mirror the existing API as closely as possible? i.e., nest these options under `http_options`

**Comment by zcin** [comment]
> @zcin can we make this mirror the existing API as closely as possible? i.e., nest these options under `http_options`

Gotcha, so move host, port, root path, and location under a new top-level field called `http_options`?

**Comment by edoakes** [comment]
So actually let me edit this, I think we should put the HTTP-specific stuff under `http_options` but not `host` and `port`.

So let's leave `host` and `port` as-is and nest the others being added under `http_options`. This will

1. Avoid thrashing from existing API
2. Be future proof if we add other transports like grpc in the future

**Review by sihanwang41 — APPROVED** [review]

**Review by edoakes — APPROVED** [review]

**Comment by zcin** [comment]
Changes since last review: Added validator in `ServeDeploySchema` that prevents users from setting host and port in individual application configs.

**Comment by zcin** [comment]
The top level fields of `ServeDeploySchema` and `ServeInstanceDetails` are pretty 1-1, so I feel like it makes sense to make the same changes to `ServeInstanceDetails` to avoid confusion with users. Made the following changes:
image

**Comment by zcin** [comment]
@edoakes Fixed merge conflicts + tests, ready for another pass / merge!

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/controller.py [hunk 1]
```diff
@@ -714,8 +716,11 @@ def get_serve_instance_details(self) -> Dict:
         # route_prefix is set instead in each application.
         # Eventually we want to remove route_prefix from DeploymentSchema.
         return ServeInstanceDetails(
-            host=http_config.host,
-            port=http_config.port,
+            proxy_location=http_config.location,
+            http_options=HTTPOptionsSchema(
+                host=http_config.host,
+                port=http_config.port,
+            ),
```
Comment (alanwguo) [inline python/ray/serve/controller.py:723]: btw, this changed broke the frontend UI. 

The ServeInstanceDetails class and all of its fields are now public apis that power the frontend UI. They need to be changed in backwards compatible ways or the frontend has to be updated when these things change.

In this case, the frontend is not yet enabled publicly to users but this is a FYI for the future.

Comment (alanwguo) [reply python/ray/serve/controller.py:723]: We should have better automated testing once we start generating the web models from the API models.

Comment (edoakes) [reply python/ray/serve/controller.py:723]: ah yeah sorry this was an intentional breakage... would be great to have some automated compatibility testing for the request & response schemas, maybe we can get that on the plan for next release

**Pull request merged**